crates/analyzer/src/db/queries/types.rs

```diff
@@ -69,3 +69,45 @@ pub fn type_alias_type_cycle(
 
     Analysis::new(err, context.diagnostics.take().into())
 }
+
+pub fn is_zero_sized(db: &dyn AnalyzerDb, ty: types::TypeId) -> bool {
+    let typ = db.lookup_intern_type(ty);
+    match typ {
+        Type::Base(base) => matches!(base, Base::Unit),
+        Type::Array(Array { size, inner }) => size == 0 || inner.is_zero_sized(db),
+        Type::Tuple(Tuple { items }) => {
+            for item in items.iter() {
+                if !db.is_zero_sized(*item) {
+                    return false;
+                }
+            }
+            true
+        }
+        Type::String(FeString { max_size }) => max_size == 0,
+        Type::Struct(struct_id) => {
+            for field_type_id in db.struct_all_fields(struct_id).iter() {
+                // assume all type is correct.
+                let field_type = db.struct_field_type(*field_type_id).value.unwrap();
+                if !(field_type).is_zero_sized(db) {
+                    return false;
+                }
+            }
+            true
+        }
+        Type::Contract(contract_id) => {
+            for field_type_id in db.contract_all_fields(contract_id).iter() {
+                let field_type = db.contract_field_type(*field_type_id).value.unwrap();
+                if !(field_type).is_zero_sized(db) {
+                    return false;
+                }
+            }
+            true
+        }
+        Type::Generic(_) | Type::Map(_) | Type::SelfContract(_) => false,
```

Comment:
Is Generic type always have size > 0?

Reply:
It could be zero-sized or not, depending on how its used. For example:

```
event Foo<T> {
  idx val: T
}

pub fn emit_foo<T>(ctx: Context, val: T) {
  emit Foo(val)
}

emit_foo(ctx, 10) // ok
emit_foo(ctx, ())  // not ok, T is zero-sized
```

For now, it's safe to use `unreachable!()` for this case, because we don't support generic fields in structs or events yet. When we do, we'll have to determine the size of the concrete type.

(Map and SelfContract can also be `unreachable!()`, as they can't appear in a field.)

Reply:
I think better we should throw `fe error` in this case. Use `unreachable!()` maybe non sense I think.

Reply:
I don't think returning `Error` here is a good option since using errors to report/hide internal bugs is not a good practice. Entering the `Type::Generic` arm means ICE in the current state, so `unreachable!()` would be suitable here.